File: atak/ATAKMapEngine/lib/src/main/java/com/atakmap/map/layer/feature/geometry/opengl/GLGeometry.java

package com.atakmap.map.layer.feature.geometry.opengl;

import com.atakmap.annotations.DeprecatedApi;
import com.atakmap.map.layer.feature.geometry.Geometry;
import com.atakmap.map.layer.feature.geometry.GeometryCollection;
import com.atakmap.map.layer.feature.geometry.LineString;
import com.atakmap.map.layer.feature.geometry.Point;
import com.atakmap.map.layer.feature.geometry.Polygon;

/** @deprecated use the batch feature renderering framework */
@Deprecated
@DeprecatedApi(since = "4.1", forRemoval = true, removeAt = "4.4")
public abstract class GLGeometry {
    public final static int VERTICES_PROJECTED = 0;
    public final static int VERTICES_PIXEL = 1;
    public final static int VERTICES_BATCH = 2;

    protected final Geometry geometry;

    protected int verticesSrid;
    protected int pixelCoordsVersion;

    GLGeometry(Geometry geom) {
        this.geometry = geom;
    }
    
    public Geometry getSubject() {
        return this.geometry;
    }
    
    public static GLGeometry createRenderer(Geometry geom) {
        if(geom instanceof Point)
            return new GLPoint((Point)geom);
        else if(geom instanceof Polygon)
            return new GLPolygon((Polygon)geom);
        else if(geom instanceof LineString)
            return new GLLineString((LineString)geom);
        else if(geom instanceof GeometryCollection)
            return new GLGeometryCollection((GeometryCollection)geom);
        else
            throw new IllegalStateException();
    }
}

File: takengine/takengine/src/main/cpp/renderer/feature/GLGeometry.cpp
#include "renderer/feature/GLGeometry.h"

#include "feature/Point.h"
#include "feature/Polygon.h"
#include "feature/LineString.h"
#include "feature/GeometryCollection.h"

#include "renderer/feature/GLPoint.h"
#include "renderer/feature/GLPolygon.h"
#include "renderer/feature/GLLineString.h"
#include "renderer/feature/GLGeometryCollection.h"

using namespace atakmap::feature;
using namespace atakmap::renderer;
using namespace atakmap::renderer::feature;

GLGeometry::GLGeometry(Geometry *geom)
: geometry(geom), verticesSrid(0), pixelCoordsVersion(0)
{ }

GLGeometry::~GLGeometry() { }

Geometry *GLGeometry::getSubject() {
    return geometry;
}

GLGeometry *GLGeometry::createRenderer(Geometry *geom) {
    Point *pt = dynamic_cast<Point *>(geom);
    if (pt != nullptr)
        return new GLPoint(pt);
    
    Polygon *poly = dynamic_cast<Polygon *>(geom);
    if (poly != nullptr)
        return new GLPolygon(poly);

    LineString *ls = dynamic_cast<LineString *>(geom);
    if (ls != nullptr)
        return new GLLineString(ls);

    GeometryCollection *gc = dynamic_cast<GeometryCollection *>(geom);
    if (gc != nullptr)
        return new GLGeometryCollection(gc);

    throw std::invalid_argument("unknown geometry");
}
